File: bonsai_skills/src/main/java/de/unibi/citec/clf/bonsai/skills/personPerception/GetPeople.java
package de.unibi.citec.clf.bonsai.skills.personPerception;

import de.unibi.citec.clf.bonsai.core.exception.CommunicationException;
import de.unibi.citec.clf.bonsai.core.object.MemorySlotWriter;
import de.unibi.citec.clf.bonsai.core.object.Sensor;
import de.unibi.citec.clf.bonsai.engine.model.AbstractSkill;
import de.unibi.citec.clf.bonsai.engine.model.ExitStatus;
import de.unibi.citec.clf.bonsai.engine.model.ExitToken;
import de.unibi.citec.clf.bonsai.engine.model.config.ISkillConfigurator;
import de.unibi.citec.clf.bonsai.util.CoordinateSystemConverter;
import de.unibi.citec.clf.bonsai.util.CoordinateTransformer;
import de.unibi.citec.clf.btl.List;
import de.unibi.citec.clf.btl.data.geometry.Pose2D;
import de.unibi.citec.clf.btl.data.person.PersonData;
import de.unibi.citec.clf.btl.data.person.PersonDataList;

import java.io.IOException;

/**
 * Use this state get current people from sensor
 *
 * <pre>
 * options:
 *
 * slots:
 * PersonDataListSlot: [PersonDataList] [Write] -> saves the current people
 *
 * possible return states are:
 * success             -> person found
 *
 * </pre>
 *
 * @author  lruegeme
 */

public class GetPeople extends AbstractSkill {



    // used tokens
    private ExitToken tokenSuccess;

    private Sensor<PersonDataList> personSensor;

    private MemorySlotWriter<PersonDataList> currentPersonSlot;


    List<PersonData> persons;
    CoordinateTransformer tf;
    private Sensor<Pose2D> positionSensor;
    private Pose2D robotPosition;

    @Override
    public void configure(ISkillConfigurator configurator) {
        // odom -> footprint broken?
         tf = (CoordinateTransformer) configurator.getTransform();

        // request all tokens that you plan to return from other methods
        tokenSuccess = configurator.requestExitToken(ExitStatus.SUCCESS());
        positionSensor = configurator.getSensor("PositionSensor", Pose2D.class);

        personSensor = configurator.getSensor("PersonSensor", PersonDataList.class);
        currentPersonSlot = configurator.getWriteSlot("PersonDataListSlot", PersonDataList.class);

    }

    @Override
    public boolean init() {
        return true;
    }

    @Override
    public ExitToken execute() {

        try {
            persons = personSensor.readLast(200);
            robotPosition = positionSensor.readLast(200);
        } catch (IOException | InterruptedException ex) {
            logger.error("Exception while retrieving stuff", ex);
            return ExitToken.fatal();
        }


        if(!persons.isEmpty() && !persons.get(0).isInBaseFrame()) {
            for (PersonData p : persons) {
                p.setPosition(getLocalPosition(p.getPosition()));
            }

        }

        PersonDataList list = new PersonDataList(persons);

        try {
            currentPersonSlot.memorize(list);
            return tokenSuccess;
        } catch (CommunicationException ex) {
            logger.fatal(
                    "Exception while storing current Person in memory!", ex);
            return ExitToken.fatal();
        }

    }

    private Pose2D getLocalPosition(Pose2D position) {
        if(position.getFrameId().equals(Pose2D.ReferenceFrame.LOCAL.getFrameName())) {
            return position;
        } else {
            return CoordinateSystemConverter.globalToLocal(position, robotPosition);
        }
    }

    @Override
    public ExitToken end(ExitToken curToken) {
        return curToken;
    }
}

File: bonsai_skills/src/main/java/de/unibi/citec/clf/bonsai/skills/personPerception/GetPeople.kt
package de.unibi.citec.clf.bonsai.skills.personPerception

import de.unibi.citec.clf.bonsai.core.exception.CommunicationException
import de.unibi.citec.clf.bonsai.core.`object`.MemorySlotWriter
import de.unibi.citec.clf.bonsai.core.`object`.Sensor
import de.unibi.citec.clf.bonsai.engine.model.AbstractSkill
import de.unibi.citec.clf.bonsai.engine.model.ExitStatus
import de.unibi.citec.clf.bonsai.engine.model.ExitToken
import de.unibi.citec.clf.bonsai.engine.model.config.ISkillConfigurator
import de.unibi.citec.clf.bonsai.util.CoordinateSystemConverter
import de.unibi.citec.clf.btl.List
import de.unibi.citec.clf.btl.data.geometry.Pose2D
import de.unibi.citec.clf.btl.data.person.PersonData
import de.unibi.citec.clf.btl.data.person.PersonDataList
import java.io.IOException

/**
 * <p>
 * Gets all persons currently detected by the robot from the sensor.
 * </p>
 *
 * <pre>
 *
 * Slots:
 * PersonDataListSlot:     [PersonDataList] [Write]
 * -> Saves the current people to this slot
 *
 * ExitTokens:
 * success:                Persons have been retrieved successfully
 *
 * </pre>
 *
 * @author  lruegeme
 */
class GetPeople : AbstractSkill() {
    // used tokens
    private var tokenSuccess: ExitToken? = null

    private var personSensor: Sensor<PersonDataList>? = null

    private var currentPersonSlot: MemorySlotWriter<PersonDataList?>? = null


    lateinit var persons: List<PersonData>
    private var positionSensor: Sensor<Pose2D?>? = null
    private var robotPosition: Pose2D? = null

    override fun configure(configurator: ISkillConfigurator) {
        tokenSuccess = configurator.requestExitToken(ExitStatus.SUCCESS())
        positionSensor = configurator.getSensor("PositionSensor", Pose2D::class.java)
        personSensor = configurator.getSensor("PersonSensor", PersonDataList::class.java)
        currentPersonSlot = configurator.getWriteSlot("PersonDataListSlot", PersonDataList::class.java)
    }

    override fun init(): Boolean {
        return true
    }

    override fun execute(): ExitToken? {
        try {
            persons = personSensor!!.readLast(200)
            robotPosition = positionSensor!!.readLast(200)
        } catch (ex: IOException) {
            logger.error("Exception while retrieving stuff", ex)
            return ExitToken.fatal()
        } catch (ex: InterruptedException) {
            logger.error("Exception while retrieving stuff", ex)
            return ExitToken.fatal()
        }

        logger.debug("Retrieved ${persons.size} persons from sensor")

        if (!persons.isEmpty() && !persons[0].isInBaseFrame) {
            for (p in persons) {
                p.position = getLocalPosition(p.position)
            }
        }

        val list = PersonDataList(persons)

        try {
            currentPersonSlot?.memorize(list)
            return tokenSuccess
        } catch (ex: CommunicationException) {
            logger.fatal("Exception while storing current people in memory!", ex)
            return ExitToken.fatal()
        }
    }

    private fun getLocalPosition(position: Pose2D): Pose2D {
        return if (position.getFrameId() == Pose2D.ReferenceFrame.LOCAL.frameName) {
            position
        } else {
            CoordinateSystemConverter.globalToLocal(position, robotPosition)
        }
    }

    override fun end(curToken: ExitToken?): ExitToken? {
        return curToken
    }
}
